### ISSUE TITLE
addArchivedFileSet to ZipArchiver does not work with tar archives

### ISSUE THREAD

**Issue by plamentotev** [codehaus-plexus/plexus-archiver#10]
When I try to add the content of tar file to zip file (using `addArchivedFileSet`) I got an exception. I created a test class that reproduces the problem:

```java
package org.codehaus.plexus.archiver;

import java.io.File;

import org.codehaus.plexus.PlexusTestCase;
import org.codehaus.plexus.archiver.tar.TarArchiver;
import org.codehaus.plexus.archiver.tar.TarLongFileMode;
import org.codehaus.plexus.archiver.util.DefaultArchivedFileSet;
import org.codehaus.plexus.archiver.zip.ZipArchiver;
import org.codehaus.plexus.logging.Logger;

public class AddArchivedFileSetTest
    extends PlexusTestCase
{
    public void setUp()
        throws Exception
    {
        super.setUp();
        getContainer().getLoggerManager().setThreshold( Logger.LEVEL_DEBUG );
    }

    public void testTarToZip()
        throws Exception
    {
        TarArchiver tarArchiver = (TarArchiver) lookup( Archiver.ROLE, "tar" );
        tarArchiver.setLongfile(TarLongFileMode.posix );

        File testTarArchive = getTestFile( "target/output/archived-file-set/input.tar" );
        for (int i = 0; i < 100; i++) {
            tarArchiver.addFile( getTestFile( "src/test/resources/manifests/manifest1.mf" ), "file" + i );
        }
        tarArchiver.setDestFile( testTarArchive );
        tarArchiver.createArchive();

        ZipArchiver zipArchiver = (ZipArchiver) lookup( Archiver.ROLE, "zip");
        zipArchiver.setDestFile( getTestFile("target/output/archived-file-set/archive.zip") );
        zipArchiver.addArchivedFileSet( new DefaultArchivedFileSet( testTarArchive ) );
        zipArchiver.createArchive();
    }
}
```

The exception is:
```
org.codehaus.plexus.archiver.ArchiverException: Problem creating zip: Execution exception (and the archive is probably corrupt but I could not delete it)
	at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:982)
	at org.codehaus.plexus.archiver.AddArchivedFileSetTest.testTarToZip(AddArchivedFileSetTest.java:38)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at junit.framework.TestCase.runBare(TestCase.java:141)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:675)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
Caused by: java.io.IOException: Execution exception
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver.close(AbstractZipArchiver.java:799)
	at org.codehaus.plexus.archiver.AbstractArchiver.createArchive(AbstractArchiver.java:969)
	... 20 more
Caused by: java.lang.RuntimeException: java.io.IOException: Error detected parsing the header
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$1.get(AbstractZipArchiver.java:491)
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$2.get(AbstractZipArchiver.java:599)
	at org.apache.commons.compress.archivers.zip.ZipArchiveEntryRequest.getPayloadStream(ZipArchiveEntryRequest.java:62)
	at org.apache.commons.compress.archivers.zip.ScatterZipOutputStream.addArchiveEntry(ScatterZipOutputStream.java:95)
	at org.apache.commons.compress.archivers.zip.ParallelScatterZipCreator$2.call(ParallelScatterZipCreator.java:175)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Error detected parsing the header
	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.getNextTarEntry(TarArchiveInputStream.java:280)
	at org.codehaus.plexus.archiver.tar.TarFile.findEntry(TarFile.java:187)
	at org.codehaus.plexus.archiver.tar.TarFile.getInputStream(TarFile.java:162)
	at org.codehaus.plexus.archiver.tar.TarFile.getInputStream(TarFile.java:134)
	at org.codehaus.plexus.archiver.tar.TarResource.getContents(TarResource.java:64)
	at org.codehaus.plexus.components.io.resources.Deferred.getContents(Deferred.java:60)
	at org.codehaus.plexus.components.io.resources.proxy.ResourceInvocationHandler.invoke(ResourceInvocationHandler.java:62)
	at com.sun.proxy.$Proxy3.getContents(Unknown Source)
	at org.codehaus.plexus.archiver.ArchiveEntry.getInputStream(ArchiveEntry.java:126)
	at org.codehaus.plexus.archiver.zip.AbstractZipArchiver$1.get(AbstractZipArchiver.java:487)
	... 8 more
Caused by: java.lang.IllegalArgumentException: Invalid byte 32 at offset 6 in '100644 00000' len=12
	at org.apache.commons.compress.archivers.tar.TarUtils.parseOctal(TarUtils.java:137)
	at org.apache.commons.compress.archivers.tar.TarUtils.parseOctalOrBinary(TarUtils.java:168)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.parseTarHeader(TarArchiveEntry.java:1013)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.parseTarHeader(TarArchiveEntry.java:996)
	at org.apache.commons.compress.archivers.tar.TarArchiveEntry.<init>(TarArchiveEntry.java:326)
	at org.apache.commons.compress.archivers.tar.TarArchiveInputStream.getNextTarEntry(TarArchiveInputStream.java:278)
	... 17 more
```

It's interesting that every time the result of the execution (the number of files processed before the exception is thrown) is different. That makes me wonder if this is related to some concurrent access to the tar file. I tried to hard code the number of threads to 1 (`zOut = new ConcurrentJarCreator(1)`) but it does not help.

**Comment by plamentotev** [comment]
So I dig a little bit further and to me looks like it's really a matter of concurrent access.  Even if the `ConcurrentJarCreator` uses single thread, the thread that compresses the archive entry is different from the one that calls `ParallelScatterZipCreator.addArchiveEntry` so even with `new ConcurrentJarCreator(1)`there are two threads that access the tar file. The one that compress the zip file and one that submits the entries to the `ConcurrentJarCreator`.

My particular use case  is that I want to distribute my maven project as zip archive but I do have third party dependency that is distributed as tar.gz. I want to distribute this dependency as files not as archive inside my archive. How I achieve this is to unpack the tar file and pack it inside the zip file. But I want to be able to use the `dependencySet` in my assembly descriptor (together with the unpack option). This way I'll skip the unpacking step. I'm not sure if this will speed up the build (I hope so - after all I'm not writing intermediate file on the disk) but the main advantage is that this way the build is multi-platform. I could pack the archives for *nix systems under windows. If I use the extra unpack step I'll lose the symlinks(as a matter of fact I could preserve the symlinks but 1. the builder must have admin privileges to create symlinks under windows and 2. the paths inside the links are with wrong separator - `\` instead of '/'), not to mention the file mode.

There is an issue in plexus archiver that when you use  `addArchivedFileSet` the resulting archive will contain zero byte files instead of the symlinks from the source archive. I have fixed that and I'll probably open merge request once I write unit tests and clean the code.

Back to the issue. I was hoping that there is some simple solution like to set the number of threads used for `ConcurrentJarCreator` to one when any of the resources added to the archive requires synchronous access(like addition of tar entries). After all it's better to archive the files slower than to not archive them at all (or even worse to archive successfully some garbage files instead of the original files). But unfortunately as I found out it's not that simple. Any ideas on this one?

**Comment by krosenvold** [comment]
Your change in plexus-io was the logical thing to do. I merged that and released plexus-io 2.7. I also updated plexus-archiver to add the missing methods.

I have not yet had the time to look at how to actually *use* the new method to restrict usage of a collection to a single thread. I am working on other releases these days, and it's going to be a little while before I get to this. If you want to have a go at fixing it, I'll definitely review it - quickly :)
